### PR TITLE
WT-10708 Add option to dump a window of records to WT util

### PR DESCRIPTION
**Pull request by marcbutler**
When dumping a record specified by a key, this option allows user to include up to N records either side of the record specified by the key.
```
wt dump -k k1 -w 2
```
Attempt to dump up to 2 records before and after k1.

### CODE REVIEW

File: test/suite/test_dump03.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+#
+# [TEST_TAGS]
+# wt_util
+# [END_TAGS]
+
+import wttest
+from suite_subprocess import suite_subprocess
+from wtscenario import make_scenarios
+
+# test_dump02.py
```
Comment (etienneptl) [inline test/suite/test_dump03.py:37]:
```suggestion
# test_dump03.py
```

File: test/suite/test_dump03.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+# Test the dump utility to find different keys.
```
Comment (etienneptl) [inline test/suite/test_dump03.py:38]:
```suggestion
# Test the window option of the dump utility.
```

File: test/suite/test_dump03.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+        cursor = self.session.open_cursor(uri, None, None)
+        for i in range(1, n_rows):
+            cursor[self.gen_key(i)] = self.gen_value(i)
+        cursor.close()
+
+    def test_window(self):
+        self.session.create(self.uri, self.table_format)
+        self.populate_table(self.uri, self.n_rows)
+
+        # The nearest key should be found.
```
Comment (etienneptl) [inline test/suite/test_dump03.py:82]: Since `-n` is not given, the exact key should be found.

**Review by etienneptl — APPROVED** [review]
LGTM! Nice functionality.

File: src/utilities/util_dump.c [hunk 4]
```diff
@@ -663,67 +674,123 @@ dump_record(WT_CURSOR *cursor, const char *key, bool reverse, bool search_near,
...
-                return (util_err(session, EIO, NULL));
+        /* Calculate the maximum window size based on how far it was possible to back up in the
+         * window above.
+         *
+         * If it was not possible to backup (n==0) it is then necessary to include the searched for
+         * record in the total window size.
+         */
+        total_window = ((n == 0) ? 1 : n) + window;
+
+        for (n = 0; n < total_window; n++) {
```
Comment (luke-pearson) [inline src/utilities/util_dump.c:748]: How are we preventing ourselves from dumping the key/value pairs twice if we are walking from point we walked back to?

File: src/utilities/util_dump.c [hunk 5]
```diff
@@ -663,67 +674,123 @@ dump_record(WT_CURSOR *cursor, const char *key, bool reverse, bool search_near,
...
+        return (util_cerr(cursor, "search_near", ret));
+
+    /*
+     * If a key has been found but search near is disabled and there is no exact match, the
+     * requested key is missing.
+     */
+    if (ret == 0 && !search_near && exact != 0)
+        return (WT_NOTFOUND);
+
+    if (window <= 0)
```
Comment (luke-pearson) [inline src/utilities/util_dump.c:725]: Given that window is a uint64_t I don't think you need the `<`.

File: src/utilities/util_dump.c [hunk 6]
```diff
@@ -663,67 +674,123 @@ dump_record(WT_CURSOR *cursor, const char *key, bool reverse, bool search_near,
...
+                    break;
+                return (util_cerr(cursor, "cursor", ret));
+            }
         }
 
-        if (ret == 0) {
-            if (fprintf(fp, "%s%s%s%s%s%s", json && once ? "," : "", prefix, current_key, infix,
-                  value, suffix) < 0)
-                return (util_err(session, EIO, NULL));
+        /* Calculate the maximum window size based on how far it was possible to back up in the
```
Comment (luke-pearson) [inline src/utilities/util_dump.c:740]: Nit: Multiline comment formatting means that the text needs to start on the second line. e.g.
/*
 * Text here.
 */

File: src/utilities/util_dump.c [hunk 7]
```diff
@@ -663,67 +674,123 @@ dump_record(WT_CURSOR *cursor, const char *key, bool reverse, bool search_near,
...
+
+    session = cursor->session;
+    once = false;
+    exact = 0;
+
+    WT_ASSERT((WT_SESSION_IMPL *)session, key != NULL);
+
+    current_key = key;
+    cursor->set_key(cursor, current_key);
+    ret = cursor->search_near(cursor, &exact);
```
Comment (luke-pearson) [inline src/utilities/util_dump.c:713]: My preference would be to break the walking into its own function with a function pointer passed to determine direction. We could save the key we found to restart the walk from that point.

File: src/utilities/util_dump.c [hunk 8]
```diff
@@ -663,67 +674,123 @@ dump_record(WT_CURSOR *cursor, const char *key, bool reverse, bool search_near,
...
+
+    if (ret != 0 && ret != WT_NOTFOUND)
+        return (util_cerr(cursor, "search_near", ret));
+
+    /*
+     * If a key has been found but search near is disabled and there is no exact match, the
```
Comment (lukech) [inline src/utilities/util_dump.c:719]: Suggest rewording to "search_near is not enabled" to hopefully make it more intuitive.

File: src/utilities/util_dump.c [hunk 9]
```diff
@@ -663,67 +674,123 @@ dump_record(WT_CURSOR *cursor, const char *key, bool reverse, bool search_near,
...
+    exact = 0;
+
+    WT_ASSERT((WT_SESSION_IMPL *)session, key != NULL);
+
+    current_key = key;
+    cursor->set_key(cursor, current_key);
+    ret = cursor->search_near(cursor, &exact);
+
+    if (ret != 0 && ret != WT_NOTFOUND)
+        return (util_cerr(cursor, "search_near", ret));
```
Comment (lukech) [inline src/utilities/util_dump.c:716]: Following this if clause, if `ret == WT_NOTFOUND` (meaning no match is found), should we return WT_NOTFOUND immediately?

File: src/utilities/util_dump.c [hunk 10]
```diff
@@ -663,67 +674,123 @@ dump_record(WT_CURSOR *cursor, const char *key, bool reverse, bool search_near,
...
-            if (fprintf(fp, "%s%s%s%s%s%s", json && once ? "," : "", prefix, current_key, infix,
-                  value, suffix) < 0)
-                return (util_err(session, EIO, NULL));
+        /* Calculate the maximum window size based on how far it was possible to back up in the
+         * window above.
+         *
+         * If it was not possible to backup (n==0) it is then necessary to include the searched for
+         * record in the total window size.
+         */
+        total_window = ((n == 0) ? 1 : n) + window;
```
Comment (lukech) [inline src/utilities/util_dump.c:746]: Should the `total_window` be the value of "n + 1 + windows" (1 being the searched-for-record), no matter if n equals 0 or not?

File: test/suite/test_dump03.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python
```
Comment (lukech) [inline test/suite/test_dump03.py:1]: Consider merging this new test file into the existing test_dump02.py and covering all the permutations of the 3 new command args "-k", "-n", and "-w".

File: src/utilities/util_dump.c (hunk 7 above)
Comment (marcbutler) [reply src/utilities/util_dump.c:713]: If the printing is done in a separate function is there any issue with just passing the currently positioned cursor down to the printing function?

File: src/utilities/util_dump.c (hunk 10 above)
Comment (marcbutler) [reply src/utilities/util_dump.c:746]: No, but it's confusing. The top loop uses `<=` so if the "left hand side" of the window is > 0, `n` will be `window+1`. However if you are the beginning of the file and n=0 it becomes necessary to include the "found" key.
I'll see if I can get it to make more sense.

File: test/suite/test_dump03.py (hunk 11 above)
Comment (marcbutler) [reply test/suite/test_dump03.py:1]: I'm not against it, but time wise not now.

File: src/utilities/util_dump.c (hunk 10 above)
Comment (lukech) [reply src/utilities/util_dump.c:746]: Thanks, got it.

File: src/utilities/util_dump.c (hunk 9 above)
Comment (marcbutler) [reply src/utilities/util_dump.c:716]: I don't know that case is handled below, and implies that `WT_NOTFOUND` might be returned when the specific key is not found. It's not documented in the public API.
This code was merged in from WT-10709 perhaps @etienneptl knows.

File: src/utilities/util_dump.c (hunk 6 above)
Comment (marcbutler) [reply src/utilities/util_dump.c]: corrected

File: src/utilities/util_dump.c (hunk 7 above)
Comment (marcbutler) [reply src/utilities/util_dump.c:713]: Due to time constraints I would like to defer any such refactoring.

File: src/utilities/util_dump.c (hunk 8 above)
Comment (marcbutler) [reply src/utilities/util_dump.c]: Simplified logic and comment.

File: src/utilities/util_dump.c (hunk 9 above)
Comment (marcbutler) [reply src/utilities/util_dump.c]: Simplified

File: src/utilities/util_dump.c (hunk 5 above)
Comment (marcbutler) [reply src/utilities/util_dump.c]: changed to ==

**Comment by marcbutler** [comment]
I'm merging this without full resolution of comments to get things moving for @etienneptl. Happy to revisit code and design in WT-10711 which will be the amalgamation of all the teams work.

File: src/utilities/util_dump.c (hunk 4 above)
Comment (luke-pearson) [reply src/utilities/util_dump.c:746]: Ah I misread. Gotcha.
